### DISCUSSION TITLE
Best way to mock external modules

### DISCUSSION THREAD

**Discussion by christian-bromann** [vitest-dev/vitest#1317] [Q&A] [answered]
Hey,

does anyone has a good example how to mock out external modules? For example I want to mock out the dependency of my module to `got`. I would create a `__mocks__/got.ts` file and set `vi.mock('got')` in my test file. Now in order to validate if `got` was called correctly I am currently doing this:

```ts
import * as gotOrig from 'got'
const got = gotOrig.default as any as MockedFunction<any>
```
So I am able to assert this:
```ts
expect(got).toBeCalledTimes(1)
```

However if I like to make specific assertion on parameters I get a lot of typing issues, e.g.:

```ts
expect(got.mock.calls[0][1].json).toEqual({ foo: 'baz' })
```

Would cause a type issue: `Object is of type 'unknown'.`, probably `MockedFunction` received an `any` as type parameter but I was also not able to pass in any got type that would make it happy.

__Question__: is there a better recommended way to do this?
__Suggestion__: loosen up the type from `unknown` to `any` for these mock calls so I don't have to do `(got.mock.calls[0][1] as any).json` everywhere

**Comment by sheremet-va** [comment, marked as answer]
We have `vi.mocked` typed utility, so you wouldn't do `as any as MockedFunction`:
```ts
vi.mocked(got).mock.calls
```
Also we recommend using `toHaveBeenCalledWith` matcher.
You can also use asymmetric matchers:
```ts
expect(got).toHaveBeenCalledWith(expect.any(), expect.objectContaining({ json: { foo: 'baz' } }))
```

**Comment by christian-bromann** [comment]
Thanks for the fast response!

It seems `expect.any()` does require a constructor:

```ts
(method) Vi.ExpectStatic.any(constructor: unknown): any
Expected 1 arguments, but got 0.ts(2554)
index.d.ts(1595, 17): An argument for 'constructor' was not provided.
``` 

Should `constructor` be optional then?

**Reply by christian-bromann** [reply]
This would work though:

```ts
expect(got).toHaveBeenCalledWith(expect.any(() => {}), expect.objectContaining({ foo: 'baz' }))
```

**Reply by sheremet-va** [reply]
we have `expect.anything()` if you don't care for what's inside. if you care for what type it is, you can use `expect.any(Function/ClassName/Object)`

**Reply by christian-bromann** [reply]
Thanks for the reply!
